### ISSUE TITLE
Audit logs

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#2203]
APIserver should generate audit logs of accesses.
Should these be just HTTP server logs, or should they have more structure?

Things to log in a structured audit log:
 -  identity associated with client, from e.g. tokenfile.Tokenfile
 -  time
 -  which apiserver if multiple ones
 - sourceIP?
 - which authorization policy matched (e.g. abac.Policy)
 - request attributes used for authorization (e.g. authorizer.Attributes)

References
  - #443

**Comment by erictune** [comment]
@preillyme mentioned a need for audit logging in his work. 

 I think most of the audit log handling should be external to the apiserver, so that various system integrators can build their own solutions.  But there needs to be a minimal API in the apiserver.  I will write up some more thoughts in a separate issue or PR.

@smarterclayton @derekwaynecarr @pmorie  can one of you point anyone at Red Hat who is interested to this issue?

**Comment by erictune** [comment]
@cjcullen @roberthbailey We talked about audit yesterday.

**Comment by erictune** [comment]
Which of the following events does an audit log need (each additional type implies another hook in the code, and probably more complexity to a pluggable audit logging interface.)

- failed authentication
- failed request authorization plugin check.
- blocked by admission controller
- blocked at quota stage
- error/timeout writing to storage layer.
- successful HTTP operation
- entire content of posted/modified object, or just name of object?
- content of watches?

 the audit log need to record both attempted and successful actions?



It could record failed A slight disadvantage is that some requests may be rejected after the authorization step, so the audit log would hold attempted but not actually completed actions, and it would not know which actions completed. I'm not sure what the best practice is there for auditing. We might need another call-out from apiserver to hostedmasterserver to record completed/rejected status of requests, so that audit log can reflect outcome.
9:50 AM Yesterday•Edit•Delete

Robert BaileyLet's file a bug (if there isn't one yet) and move the discussion there.

**Comment by ghodss** [comment]
Audit logging will definitely become a big blocker for us as we onboard production load onto our clusters.

One option is to piggy back off the current event objects and enhance that with more fields and/or different types of events. It feels like there is a pretty large overlap of use cases and information between events and audit logs. That way shims can be written to watch events using all the existing api infrastructure (and even kubectl get events --watch) and convert them into log files, or directly pipe those events into splunk, elasticsearch, kafka, etc. I don't think you need a plugin or a synchronous interface that kube calls or to persist the events in kube in any way.

In general I think it's better to log everything and allow people to filter out things they don't want, and just try to log efficiently (e.g. diffs vs. whole contents when possible, etc.).

**Comment by smarterclayton** [comment]
Agree on logging everything - you can't go back and add logging for the
error that happens 1 time in a million in a production cluster.

If events weren't such a bottleneck today, I'd be more comfortable with
taking our first audit steps there.  Audit -> ES, Splunk, Kafka is the
right solution.

**Comment by jonmoter** [comment]
Any movement on this? Having an audit trail of mutations in the system (like creating pods, deployments, etc.) is a requirement for compliance reasons. In particular, keeping track of which users executed which request.

I've looked at /var/log/kube-apiserver.log on the masters, and that has a nice list of all the API calls done, but I don't see anything identifying the user that made the request.

All the stuff Eric mentioned in the original story would be great as well. I'm just bringing up the bare minimum of the info we need for audit purposes.

**Comment by roberthbailey** [comment]
@jonmoter until we have fine grained users (or roles), an audit log wouldn't be very useful because all client access is done via the same credential (client cert that is shared among all users, which all have the same privilege level). @erictune can you cross link the IAM issue(s)?

**Comment by jonmoter** [comment]
I was planning on creating multiple certificates tied to users with different privilege levels.

One cert for our deploy tool, another widely-available cert that has limited read-only access, and another cert that would only be readable by root on our servers, and end-users would need to do use `sudo` to access (and we log all sudo requests).

So at the very least we'd need to be able to differentiate between commands executed by our deployment application from commands executed with the other certs. It's not ideal, but we can partially solve the problem by generating multiple certs with multiple access levels.  (right?)

**Comment by roberthbailey** [comment]
Yes, if you can separate out your users into roles with different identities, then having audit logs for those roles seems like it'd be awesome. I don't think anyone has worked on this yet though. Do you want to write up a proposal?

**Comment by smarterclayton** [comment]
@soyltsh your Openshift PR may be relevant here

**Comment by soltysh** [comment]
I'll see how useful ours implementation can be, esp. now that part of it's user-related stuff are in.

**Comment by ericchiang** [comment]
We'd also be interested in this.

From conversations on #sig-auth this would be great for debugging purposesa as well. Even with small amounts of users the auth process is pretty opaque at the moment, and this is undoubtedly going to get worse as RBAC and other authenticators become more stable and get more use.

Would an initial implementation that defined a "Logger" interface similar to the authorizer[0] and authenticator[1] interfaces be reasonable?

[0] https://godoc.org/k8s.io/kubernetes/pkg/auth/authorizer#Authorizer
[1] https://godoc.org/k8s.io/kubernetes/pkg/auth/authenticator#Request

**Comment by soltysh** [comment]
I'll try to put down the PR later today and ping you all.

**Comment by jonmoter** [comment]
@soltysh, any word on your PR?

**Comment by ericchiang** [comment]
@jonmoter https://github.com/kubernetes/kubernetes/pull/27087
